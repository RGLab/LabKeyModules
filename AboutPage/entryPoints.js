/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
 */
module.exports = {
    apps: [
        // For each webpart, include an object matching
        {
            name: 'AboutPage',
            frame: 'portal',
            title: 'About Page',
            path: './src/client/AboutPage',
            domElementId: 'about-page'
        }]
};