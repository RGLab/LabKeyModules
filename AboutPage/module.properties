Author: ops@immunespace.org
Description: This module contains general information about ImmuneSpace
InfoURL: https://github.com/RGLab/LabKeyModules/tree/main/AboutPage
Label: About Page
License: Apache 2.0
LicenseURL: http://www.apache.org/licenses/LICENSE-2.0
Maintainer: ops@immunespace.org
ModuleClass: org.labkey.api.module.SimpleModule
ModuleDependencies: ISCore
Name: AboutPage
Organization: ImmuneSpace Development Team
OrganizationURL: http://www.immunespace.org
RequiredServerVersion: 14.1
SchemaVersion: 0.01
