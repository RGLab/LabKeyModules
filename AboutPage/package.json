{
  "name": "aboutpage",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "setup": "npm install",
    "build": "npm run build-prod",
    "start": "cross-env NODE_ENV=development LK_MODULE=AboutPage webpack serve --config ../webpack/dev.config.js",
    "build-prod": "npm run clean && cross-env NODE_ENV=production LK_MODULE=AboutPage webpack --config ../webpack/prod.config.js",
    "clean": "rimraf resources/web/AboutPage/gen && rimraf resources/views",
    "test": "cross-env NODE_ENV=test jest"
  },
  "dependencies": {
    "@types/jquery": "^3.5.0",
    "bootstrap": "^3.4.1",
    "d3-color": "^1.4.0",
    "font-awesome": "4.7.0",
    "jquery": "^3.5.1",
    "react-router-dom": "^5.2.0"
  },
  "devDependencies": {
    "cross-env": "5.2.1",
    "rimraf": "3.0.2",
    "webpack": "5.22.0",
    "webpack-cli": "4.5.0",
    "webpack-dev-server": "3.11.2"
  },
  "jest": {
    "globals": {
      "LABKEY": {}
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "setupFilesAfterEnv": [
      "<rootDir>/test/jest.setup.ts"
    ],
    "moduleNameMapper": {
      "\\.(scss)$": "<rootDir>/test/styleMock.ts"
    },
    "testRegex": "(\\.(test|spec))\\.(ts|tsx)$",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    }
  }
}
