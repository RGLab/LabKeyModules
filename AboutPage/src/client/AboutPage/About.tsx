import * as React from "react";

export const About = () => {
  return (
    <div id="About-content">
      <p>
        ImmuneSpace blossomed from a shared idea that generating large amounts
        collaborative data, cross-center and cross-assay, to characterize the
        status of the immune system in diverse populations under normal
        conditions and in response to various stimuli would be a beneficial
        platform. The Human Immunology Project Consortium (
        <a
          href="https://www.immuneprofiling.org/hipc/page/show?pg=home"
          target="_blank">
          HIPC
        </a>
        ) program, an organization founded by the
        <a href="https://www.niaid.nih.gov/about/dait" target="_blank">
          {" "}
          NIAID-DAIT
        </a>
        , released ImmuneSpace in January 15, 2016.
      </p>
      <p>
        Since its release, ImmuneSpace has developed into powerful data
        management and analysis engine, where datasets can be easily explored
        and analyzed using state-of-the-art computational tools. ImmuneSpace
        takes advantage of the considerable infrastructure already developed
        through another HIPC platform,
        <a href="https://immport.niaid.nih.gov" target="_blank">
          {" "}
          ImmPort
        </a>
        , which serves as a repository of data generated by investigators funded
        by DAIT. Data are submitted by each HIPC center to ImmPort using a set
        of standardized data templates. Once ImmPort has made these data public
        through routine data releases, the data are transferred to ImmuneSpace
        where integrative modeling across various data types and HIPC centers
        are enabled.
      </p>
      <p>
        ImmuneSpace provides multiple ways to interact with, visualize, and
        analyze data. Each study contains tabs to view raw data, run common
        analyses, and look at custom reports. All the analyses make use of the R
        statistical language, leveraging
        <a href="http://rforge.net/Rserve/" target="_blank">
          {" "}
          Rserve{" "}
        </a>
        to improve performance and
        <a href="http://yihui.name/knitr/" target="_blank">
          {" "}
          knitr{" "}
        </a>
        to enable full reproducibility.
      </p>
      <img
        src="/AboutPage/images/dataflow.jpeg"
        padding-top="80%"
        width="40%"
        style={{ alignSelf: "center" }}
      />
      <p></p>
      <p>
        <b>Support:</b>
      </p>
      <ul>
        <li>
          Slack: The best way to connect with the ImmuneSpace team is via the
          ImmuneSpace{" "}
          <a href="https://immunespace.herokuapp.com/">slack workspace</a>
        </li>
        <li>
          Email: You can also reach the team at <b>immunespace@gmail.com</b>
        </li>
      </ul>
    </div>
  );
};
