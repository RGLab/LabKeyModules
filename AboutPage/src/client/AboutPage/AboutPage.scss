@import "../theme/AboutPage";

#about-page-content {
  margin: 0 15px;
}
