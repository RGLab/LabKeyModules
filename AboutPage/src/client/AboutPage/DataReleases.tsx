import * as React from "react";

interface props {
  dataReleasesResults: JSX.Element | JSX.Element[];
}

export const DataReleases = React.memo<props>(
  ({ dataReleasesResults }: props) => {
    return (
      <div id="DataReleases-content">
        <table style={{ width: "100%", border: "1px solid black" }}>
          <thead>
            <tr>
              <th
                style={{
                  width: "15%",
                  border: "1px solid black",
                  textAlign: "center",
                  fontWeight: "bold",
                }}>
                Version
              </th>
              <th
                style={{
                  width: "15%",
                  border: "1px solid black",
                  textAlign: "center",
                  fontWeight: "bold",
                }}>
                Date
              </th>
              <th
                style={{
                  width: "30%",
                  border: "1px solid black",
                  textAlign: "center",
                  fontWeight: "bold",
                }}>
                Affected Studies
              </th>
              <th
                style={{
                  width: "40%",
                  border: "1px solid black",
                  textAlign: "center",
                  fontWeight: "bold",
                }}>
                Description
              </th>
            </tr>
          </thead>
          <tbody>{dataReleasesResults}</tbody>
        </table>
        <section className="DataReleases-footer">
          <p>
            Version numbers use the following scheme <b>x.y.z</b> where
          </p>
          <ul>
            <li>
              <b>x</b> is the version of ImmPort. The data is curated by ImmPort
              and ImmuneSpace gets updated after each offcial data release.
            </li>
            <li>
              <b>y</b> indicates a major data change. This includes loading new
              studies, adding datasets to existing studies or processing data
              such as creating gene expression matrices.
            </li>
            <li>
              <b>z</b> indicates minor changes, such as reloading studies with
              minor corrections to existing assay or metadata.
            </li>
          </ul>
          <p>
            Note that this only for <b>data</b>, the development of new features
            or updates to software infrastructure are tracked separately in{" "}
            <a href="?tab=SoftwareUpdates">Software Updates</a>.
          </p>
        </section>
      </div>
    );
  }
);

DataReleases.displayName = "DataReleases";
