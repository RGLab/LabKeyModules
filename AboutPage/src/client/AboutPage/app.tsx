import * as React from "react";
import * as ReactDOM from "react-dom";

import { App } from "./AboutPage";

window.addEventListener("DOMContentLoaded", (event) => {
  ReactDOM.render(<App />, document.getElementById("about-page"));
});
