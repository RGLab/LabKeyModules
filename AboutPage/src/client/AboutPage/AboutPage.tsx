import React from "react";
import "regenerator-runtime/runtime";

// Styling imports
import "./AboutPage.scss";

import { About } from "./About";
import { DataReleases } from "./DataReleases";
import { DataStandards } from "./DataStandards";
import { SoftwareUpdates } from "./SoftwareUpdates";
import { RSessionInfo } from "./RSessionInfo";
import { ScriptLoader } from "./ScriptLoader";
import {
  MenuItem,
  Nav,
  Navbar,
  NavDropdown,
  NavItem,
  Tab,
  TabPane,
  TabContainer,
} from "react-bootstrap";
import {
  TAB_ABOUT,
  TAB_CYTOMETRY,
  TAB_DATARELEASES,
  TAB_DATASTANDARDS,
  TAB_GENEEXPRESSION,
  TAB_IMMUNERESPONSE,
  TAB_RSESSIONINFO,
  TAB_SOFTWAREUPDATES,
  tabInfo,
} from "./constants";
import { Cytometry } from "./Cytometry";
import { GeneExpression } from "./GeneExpression";
import { ImmuneResponse } from "./ImmuneResponse";

import { active } from "d3-transition";

const fetchData = (handleResults: (any) => void) => {
  let mappedData;

  LABKEY.Query.selectRows({
    schemaName: "lists",
    queryName: "Data Updates",
    columns: ["version", "date", "affected_studies", "description"],
    success: function (data) {
      mappedData = data.rows.map(function (arr, index) {
        return (
          <tr key={index} data-item={arr}>
            <td
              data-title="Version"
              style={{ textAlign: "center", border: "1px solid black" }}
            >
              {arr.version}
            </td>
            <td
              data-title="Date"
              style={{ textAlign: "center", border: "1px solid black" }}
            >
              {arr.date.slice(0, 10)}
            </td>
            <td
              data-title="Affected Studies"
              style={{ border: "1px solid black" }}
            >
              {arr.affected_studies}
            </td>
            <td data-title="Description" style={{ border: "1px solid black" }}>
              {arr.description}
            </td>
          </tr>
        );
      });
      handleResults(mappedData);
    },
  });
};

const AboutPage: React.FC = () => {
  const [divToShow, setDivToShow] = React.useState<string>("About");
  const [dataReleasesResults, setDataReleasesResults] = React.useState<string>(
    "Loading Data Releases"
  );
  const [rSessionResults, setRSessionResults] = React.useState<string>(
    "Loading R Session Info ..."
  );
  const [rSessionParsed, setRSessionParsed] = React.useState<
    DocumentFragment
  >();
  const [rScriptsLoaded, setRScriptsLoaded] = React.useState(false);

  // finds the value of "tab" parameter in url
  const getCurrentTabParam = (): string => {
    const params = new URL(`${document.location}`).searchParams;
    const tabName = params.get("tab");
    return tabName;
  };

  const defaultAciveTab = getCurrentTabParam() ?? "About";

  const [activeTab, setActiveTab] = React.useState(
    getCurrentTabParam() ?? "About"
  );

  /* 
    When a new tab is selected, append the appropriate parameter to the end of the url
    and updates the url. Creates new history entries that allows user to traverse tabs 
    using the foward and back buttons. Does not refresh the page.

    https://developer.mozilla.org/en-US/docs/Web/API/History/pushState
  */
  const changeTabParam = (newActiveTab: string) => {
    // window.history.pushState({ tab: newActiveTab }, "", `?tab=${newActiveTab}`);
    const url = new URL(`${window.location}`);
    url.searchParams.set("tab", newActiveTab);
    window.history.pushState({}, "", `${url}`);
    setActiveTab(newActiveTab);
  };

  // handles forward/back button clicks
  // https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onpopstate
  window.onpopstate = (e) => {
    setActiveTab(getCurrentTabParam() ?? "About");
  };

  // Only declare this on the first render
  const cnfReport = React.useMemo(
    () => ({
      failure: function () {
        setRSessionResults("Unknown Error within R Session Info Report");
      },
      reportId: "module:RSessionInfo/RSessionInfo.Rmd",
      success: function (result) {
        var errors = result.errors;
        var outputParams = result.outputParams;
        if (errors && errors.length > 0) {
          setRSessionResults("Error in retrieving R Session Info");
        } else if (outputParams && outputParams.length > 0) {
          var p = outputParams[0];
          setRSessionResults(p.value);
        } else {
          setRSessionResults(
            "Strange situation: there are no reported errors, but also no output to show"
          );
        }
      },
    }),
    []
  );

  const onRScriptsLoaded = React.useCallback(() => {
    setRScriptsLoaded(true);
  }, []);

  const onRScriptLoadTimeout = React.useCallback((script: string) => {
    console.error("Failed to load R script: " + script);
  }, []);

  React.useEffect(() => {
    fetchData(setDataReleasesResults);
    LABKEY.Report.execute(cnfReport);

    // append the ?tab parameter to the url on page load
    changeTabParam(defaultAciveTab);
  }, []);

  // Want to load scripts from Rmd output once and in order they have been delivered to avoid rendering issues,
  // e.g. DataTable / jQuery conflict
  React.useEffect(() => {
    const slotHtml = document
      .createRange()
      .createContextualFragment(rSessionResults);
    let scriptNodes = slotHtml.querySelectorAll("script[src]");

    if (scriptNodes.length > 0) {
      const loader = new ScriptLoader(onRScriptsLoaded, onRScriptLoadTimeout);

      scriptNodes.forEach(function (el) {
        loader.addScript(el.getAttribute("src"));
        el.parentNode.removeChild(el);
      });

      loader.load();
      setRSessionParsed(slotHtml);
    }
  }, [rSessionResults]);

  const generateChildId = React.useCallback((eventKey: any, type: any) => {
    return eventKey;
  }, []);

  const getNavbar = React.useCallback(() => {
    const items = tabInfo.map((tab, index) => {
      // Drop down item
      if (tab.subMenu && tab.subMenu.length > 0) {
        const menuItems = tab.subMenu.map((sub) => {
          return (
            <MenuItem eventKey={sub.tag} id={sub.tag} key={sub.tag}>
              {sub.text}
            </MenuItem>
          );
        });

        return (
          <NavDropdown title={tab.text} id={tab.id} key={tab.tag}>
            {menuItems}
          </NavDropdown>
        );
      }

      // Non-dropdown
      return (
        <NavItem eventKey={tab.tag} id={tab.id} key={tab.tag}>
          {tab.text}
        </NavItem>
      );
    });

    return (
      <Navbar>
        <Nav style={{ marginLeft: "0" }}>{items}</Nav>
      </Navbar>
    );
  }, []);

  const getTabContent = React.useCallback(() => {
    return (
      <Tab.Content>
        <TabPane eventKey={TAB_ABOUT}>
          <About />
        </TabPane>
        <TabPane eventKey={TAB_DATASTANDARDS}>
          <DataStandards />
        </TabPane>
        <TabPane eventKey={TAB_CYTOMETRY}>
          <Cytometry />
        </TabPane>
        <TabPane eventKey={TAB_GENEEXPRESSION}>
          <GeneExpression />
        </TabPane>
        <TabPane eventKey={TAB_IMMUNERESPONSE}>
          <ImmuneResponse />
        </TabPane>
        <TabPane eventKey={TAB_DATARELEASES}>
          <DataReleases
            dataReleasesResults={dataReleasesResults}
            setDivToShow={setDivToShow}
          />
        </TabPane>
        <TabPane eventKey={TAB_SOFTWAREUPDATES}>
          <SoftwareUpdates />
        </TabPane>
        <TabPane eventKey={TAB_RSESSIONINFO} mountOnEnter={true}>
          <RSessionInfo
            rSessionParsed={rSessionParsed}
            rScriptsLoaded={rScriptsLoaded}
          />
        </TabPane>
      </Tab.Content>
    );
  }, [dataReleasesResults, setDivToShow, rSessionParsed, rScriptsLoaded]);

  return (
    <Tab.Container
      activeKey={activeTab}
      generateChildId={generateChildId}
      onSelect={(tab) => changeTabParam(`${tab}`)}
    >
      <div>
        {getNavbar()}
        {getTabContent()}
      </div>
    </Tab.Container>
  );
};

export const App: React.FC = () => {
  const filterBanner = document.getElementById("filter-banner");
  if (filterBanner) {
    filterBanner.style.display = "none";
  }

  // Must return a React Fragment
  return <AboutPage />;
};
