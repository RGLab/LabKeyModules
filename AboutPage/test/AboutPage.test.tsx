/**
 * @jest-environment jsdom
 */
import React from "react";
import { About } from "../src/client/AboutPage/About";
import renderer from "react-test-renderer";
import { DataReleases } from "../src/client/AboutPage/DataReleases";

describe("tests AboutPage pages renders correctly", () => {
  test("test About renders correctly", () => {
    const component = renderer.create(<About />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
