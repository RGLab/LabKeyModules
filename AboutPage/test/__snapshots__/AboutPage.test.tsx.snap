// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests AboutPage pages renders correctly test About renders correctly 1`] = `
<div
  id="About-content"
>
  <section
    className="About-content-primary"
  >
    <div
      className="About-content-text"
    >
      <h1>
        About ImmuneSpace
      </h1>
      <p>
        ImmuneSpace blossomed from a shared idea that generating large amounts collaborative data, cross-center and cross-assay, to characterize the status of the immune system in diverse populations under normal conditions and in response to various stimuli would be a beneficial platform. The Human Immunology Project Consortium (
        <a
          href="https://www.immuneprofiling.org/hipc/page/show?pg=home"
          rel="noopener noreferrer"
          target="_blank"
        >
          HIPC
        </a>
        ) program, an organization founded by the
        <a
          href="https://www.niaid.nih.gov/about/dait"
          rel="noopener noreferrer"
          target="_blank"
        >
           
          NIAID-DAIT
        </a>
        , released ImmuneSpace in January 15, 2016.
      </p>
      <p>
        Since its release, ImmuneSpace has developed into powerful data management and analysis engine, where datasets can be easily explored and analyzed using state-of-the-art computational tools. ImmuneSpace takes advantage of the considerable infrastructure already developed through another HIPC platform,
        <a
          href="https://immport.niaid.nih.gov"
          rel="noopener noreferrer"
          target="_blank"
        >
           
          ImmPort
        </a>
        , which serves as a repository of data generated by investigators funded by DAIT. Data are submitted by each HIPC center to ImmPort using a set of standardized data templates. Once ImmPort has made these data public through routine data releases, the data are transferred to ImmuneSpace where integrative modeling across various data types and HIPC centers are enabled.
      </p>
      <p>
        ImmuneSpace provides multiple ways to interact with, visualize, and analyze data. Each study contains tabs to view raw data, run common analyses, and look at custom reports. All the analyses make use of the R statistical language, leveraging
        <a
          href="http://rforge.net/Rserve/"
          rel="noopener noreferrer"
          target="_blank"
        >
           
          Rserve
           
        </a>
        to improve performance and
        <a
          href="http://yihui.name/knitr/"
          rel="noopener noreferrer"
          target="_blank"
        >
           
          knitr
           
        </a>
        to enable full reproducibility.
      </p>
    </div>
    <div
      className="About-content-img"
    >
      <img
        alt="dataflow"
        src="/AboutPage/images/flow.png"
      />
    </div>
  </section>
  <section
    className="About-content-support"
  >
    <p>
      <b>
        Support:
      </b>
    </p>
    <p>
      Connect with us via Slack
       
      <a
        href="https://immunespace.herokuapp.com/"
      >
        slack workspace
      </a>
    </p>
    <p>
      You can also reach the team at
       
      <a
        href="mailto:ops@immunespace.org"
      >
        ops@immunespace.org
      </a>
    </p>
  </section>
</div>
`;
